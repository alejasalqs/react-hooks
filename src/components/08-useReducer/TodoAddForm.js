import React from "react";
import { useForm } from "../../Hooks/useForm";

export const TodoAddForm = ({ handleNewTodo }) => {
  const [{ description }, handleInputChange, reset] = useForm({
    description: "",
  });

  const handleSubmit = (e) => {
    e.preventDefault();
    if (description.trim().length <= 1) {
      return;
    }
    const newTodo = {
      id: new Date().getTime(),
      description: description,
      done: false,
    };

    handleNewTodo(newTodo);

    reset();
  };
  return (
    <>
      <h4>Agregar</h4>
      <hr />

      <form onSubmit={handleSubmit}>
        <input
          type="text"
          name="description"
          placeholder="description"
          autoComplete="off"
          className="form-control"
          onChange={handleInputChange}
          value={description}
        />
        <button
          type="submit"
          className="btn btn-outline-primary mt-1 btn-block"
        >
          Agregar
        </button>
      </form>
    </>
  );
};
