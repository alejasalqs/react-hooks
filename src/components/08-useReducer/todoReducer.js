export const todoReducer = (state = [], action) => {
  switch (action.type) {
    case "add": // Agrega un nuevo registro utilizando el espread operator
      return [...state, action.payload];
    case "delete": // filter regresa un nuevo arreglo
      return state.filter((todo) => todo.id !== action.payload); // En este caso el payload es el id
    case "toggle": //version corta del toggle
      return state.map((todo) =>
        todo.id === action.payload ? { ...todo, done: !todo.done } : todo
      );
    case "toggle-old": // version larga del toggle
      return state.map((todo) => {
        if (todo.id === action.payload) {
          return {
            ...todo,
            done: !todo.done,
          };
        } else {
          // Es importante retornar algo siempre en este caso
          return todo;
        }
      });

    default:
      return state; // Es mejor siempre definir el defaul
  }
};
