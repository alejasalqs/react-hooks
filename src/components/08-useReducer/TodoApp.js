import React, { useEffect, useReducer } from "react";
import "./styles.css";
import { TodoAddForm } from "./TodoAddForm";
import { TodoList } from "./TodoList";
import { todoReducer } from "./todoReducer";
// Un useReducer es una alternativa al useState
// Es mejor para estados muy complejos

const init = () => {
  // Si el primer valor es null retorna un objeto vacio
  return JSON.parse(localStorage.getItem("todos")) || []; //Lo que retornar esta funcion serÃ¡ el initialState
};

export const TodoApp = () => {
  //1 argumento: funcion del reducer
  // 2 argumento: estado inicial
  // 3 argumento: Funcion para iniciar states si es procesado o realiza varias acciones
  // dispatch: dispara las acciones al reducer
  const [todos, dispatch] = useReducer(todoReducer, [], init);

  useEffect(() => {
    // si los todos cambian grabamos en el localStorage
    localStorage.setItem("todos", JSON.stringify(todos));
  }, [todos]);

  const handleNewTodo = (newTodo) => {
    dispatch({
      type: "add",
      payload: newTodo,
    });
  };
  const handleDelete = (todoId) => {
    const action = {
      // creamos la accion
      type: "delete",
      payload: todoId,
    };

    // hacemos dispatch del reducer
    dispatch(action);
  };

  const handleToggle = (todoId) => {
    dispatch({
      // otra manera de hacer el dispatch
      type: "toggle",
      payload: todoId,
    });
  };

  /*const handleSubmit = (e) => {
    e.preventDefault();
    if (description.trim().length <= 1) {
      return;
    }

    const newTodo = {
      id: new Date().getTime(),
      description: description,
      done: false,
    };

    const action = {
      type: "add",
      payload: newTodo,
    };

    // Esta funcion es la encargada de llamar el reducer
    // Se puede mandar como referencia a un componente hijo y siempre va a saber
    // a que reducer pertenece porque se pasan por referencia en JS
    dispatch(action);

    reset();
  };*/

  return (
    <>
      <h1>Todo APP ({todos.length})</h1>
      <hr />

      <div className="row">
        <div className="col-md-6">
          <TodoList
            todos={todos}
            handleToggle={handleToggle}
            handleDelete={handleDelete}
          />
        </div>
        <div className="col-md-6">
          <TodoAddForm handleNewTodo={handleNewTodo} />
        </div>
      </div>
    </>
  );
};
