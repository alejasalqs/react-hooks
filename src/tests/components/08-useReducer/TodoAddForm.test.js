import React from "react";
import Adapter from "enzyme-adapter-react-16";
const { shallow, configure } = require("enzyme");
import { TodoAddForm } from "../../../components/08-useReducer/TodoAddForm";

configure({ adapter: new Adapter() });

describe("Pruebas en el componente <TodoAddForm />", () => {
  const handleNewTodo = jest.fn();
  const wrapper = shallow(<TodoAddForm handleNewTodo={handleNewTodo} />);
  test("Debe mostrarse correctamente ", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("NO debe de llamar handleAddTodo ", () => {
    const formSubmit = wrapper.find("form").prop("onSubmit"); // Esto es una funcion

    // Se tiene que enviar como parametro el e de preventDefault
    formSubmit({ preventDefault() {} });

    expect(handleNewTodo).toHaveBeenCalledTimes(0);
  });

  test("Debe de llamar la funcion handleNewTodo ", () => {
    // Valor que se va a cambiar
    const value = "Nuevo TODO de prueba";

    // Buscamos la caja de texto y cambiamos el input
    wrapper.find("input").simulate("change", {
      target: {
        value,
        name: "description",
      },
    });

    // definimos el evento
    const formSubmit = wrapper.find("form").prop("onSubmit"); // Esto es una funcion

    // Se tiene que enviar como parametro el e de preventDefault
    formSubmit({ preventDefault() {} });

    expect(handleNewTodo).toHaveBeenCalledTimes(1);

    expect(handleNewTodo).toHaveBeenCalledWith(expect.any(Object)); // {}
    /*expect(handleNewTodo).toHaveBeenCalledTimes({
      id: expect.any(Number),
      description: value,
      done: false,
    });*/

    expect(wrapper.find("input").prop("value")).toBe("");
  });
});
