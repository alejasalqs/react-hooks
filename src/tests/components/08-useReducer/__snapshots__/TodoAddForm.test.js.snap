// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <TodoAddForm /> Debe mostrarse correctamente  1`] = `ShallowWrapper {}`;
